### ISSUE TITLE
Bug in the json code?

### ISSUE THREAD

**Issue by jkitchin** [melpa/melpa#3548]
I am having problems running the make json target in the melpa repo.

16:09 $ make json
 • Building html/archive.json ...
emacs --no-site-file --batch -l package-build.el --eval '(let ((package-build-stable nil) (package-build-archive-dir (expand-file-name "./packages/" package-build--this-dir))) (package-build-archive-alist-as-json "html/archive.json"))'
Wrong type argument: listp, "jkitchin@andrew.cmu.edu"
make: **\* [html/archive.json] Error 255

It seems that :maintainer in the alist should be a cons cell, but it is not:

`#+BEGIN_SRC emacs-lisp :results code
(package-build--archive-alist-for-json)
# +END_SRC
# +RESULTS:
# +BEGIN_SRC emacs-lisp

(:kitchingroup-57
 (:ver
  (20160130 1223)
  :deps
  (:cappa
   (0))
  :desc "preprint http://dx.doi.org/10.1021/acscatal.5b00538." :type tar :props
  ((:authors
    ("John Kitchin" . "jkitchin@andrew.cmu.edu"))
   (:maintainer "John Kitchin" . "jkitchin@andrew.cmu.edu")))`

Any suggestions on how to track this down? I can't figure out where maintainer is even getting stuck in.

thanks,

**Comment by purcell** [comment]
Which Emacs version do you have?

**Comment by jkitchin** [comment]
GNU Emacs 25.1.50.1

**Comment by purcell** [comment]
That actually appears to be a bug in the recent Emacs you're using (and I can reproduce it locally too). In an affected buffer, e.g. the contents of [this package](https://github.com/lujun9972/uimage/blob/master/uimage.el), evaluating `(package-desc-extras (package-buffer-info))` results in the bogus output:

```el
((:authors
  ("DarkSun" . "lujun9972@gmail.com"))
 (:maintainer "DarkSun" . "lujun9972@gmail.com"))
```

where the invalid `:maintainer` entry seems to have come from `package.el` upstream. Any idea about this, @Malabarba?

**Comment by Malabarba** [comment]
I can look into it, but I don't remember changing that recently. Officially, a package is considered to have only one maintainer, and so the cdr of the :maintainer entry is a cons cell (as opposed to the cdr of :authors entry which is an alist).

**Comment by jkitchin** [comment]
Maybe the bug is really in json-encode then. This is what is actually failing, but it works fine if I make the maintainer be a list.

`Debugger entered--Lisp error: (wrong-type-argument listp "jkitchin@andrew.cmu.edu")
  mapconcat(json-encode ("John Kitchin" . "jkitchin@andrew.cmu.edu") ",")
  json-encode-array(("John Kitchin" . "jkitchin@andrew.cmu.edu"))
  json-encode-list(("John Kitchin" . "jkitchin@andrew.cmu.edu"))
  json-encode(("John Kitchin" . "jkitchin@andrew.cmu.edu"))
  #[(cons) "\303\203	\304\202\n\305	\306\n@!\307\nA!$\207" [json-encoding-pretty-print json--encoding-current-indentation cons format "%s%s: %s" "%s%s:%s" json-encode-key json-encode] 6]((:maintainer "John Kitchin" . "jkitchin@andrew.cmu.edu"))
  mapcar(#[(cons) "\303\203	\304\202\n\305	\306\n@!\307\nA!$\207" [json-encoding-pretty-print json--encoding-current-indentation cons format "%s%s: %s" "%s%s:%s" json-encode-key json-encode] 6] ((:authors ("John Kitchin" . "jkitchin@andrew.cmu.edu")) (:maintainer "John Kitchin" . "jkitchin@andrew.cmu.edu")))
  json-encode-alist(((:authors ("John Kitchin" . "jkitchin@andrew.cmu.edu")) (:maintainer "John Kitchin" . "jkitchin@andrew.cmu.edu")))
  json-encode-list(((:authors ("John Kitchin" . "jkitchin@andrew.cmu.edu")) (:maintainer "John Kitchin" . "jkitchin@andrew.cmu.edu")))
  json-encode(((:authors ("John Kitchin" . "jkitchin@andrew.cmu.edu")) (:maintainer "John Kitchin" . "jkitchin@andrew.cmu.edu")))`

**Comment by Malabarba** [comment]
Maybe json doesn't know how to encode cons cells that aren't part of an alist? 

Still, something must have changed somewhere to trigger this issue.

**Comment by jkitchin** [comment]
I think it is an error in json. Here is the evidence:

#+BEGIN_SRC emacs-lisp
(condition-case nil
    (json-encode '(:maintainer "John" . "my@email.com"))
  (error "error"))
#+END_SRC

#+RESULTS:
: error

#+BEGIN_SRC emacs-lisp
(condition-case nil
    (json-encode '(:maintainer ("John" . "my@email.com")))
  (error "error"))
#+END_SRC

#+RESULTS:
: error


#+BEGIN_SRC emacs-lisp
(condition-case nil
    (json-encode '(:maintainer (("John" . "my@email.com"))))
  (error "error"))
#+END_SRC

#+RESULTS:
: {"maintainer":{"John":"my@email.com"}}

I am not sure where to go with this?

**Comment by Malabarba** [comment]
The next would be to figure out what changed. Either json encode somehow forgot how to encode cons cells (unlikely, since I don't think the json format has anything analogue to a cons cell), or lm-maintainer used to return a different format, or package-buffer-info didn't use to include maintainer information.

**Comment by purcell** [comment]
Yeah, would try the same with the latest stable Emacs, for example.

**Comment by jkitchin** [comment]
I tried this in an older GNU Emacs 24.4.50.3 and a current GNU Emacs
25.1.50.1 from git.

In both cases his code does not work:
 (json-encode '(:maintainer "John" . "yy@emai.com"))

And this code does work:
(json-encode '(:maintainer (("John" . "yy@emai.com"))))

The problem is eventually the thing we are encoding gets passed to
json-encode-array which I think is incorrect for this.

John

-----------------------------------
Professor John Kitchin
Doherty Hall A207F
Department of Chemical Engineering
Carnegie Mellon University
Pittsburgh, PA 15213
412-268-7803
@johnkitchin
http://kitchingroup.cheme.cmu.edu

**Comment by tarsius** [comment]
Whatever the Emacs bug was, it appears to be fixed in `25.2`/`25.3`.
